### PR TITLE
Deploy dc/os with terraform

### PR DESCRIPTION
**Pull request by rimusz**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,115 +1,98 @@
...
 SSH_USER := core
-DCOS_INSTALLER_URL := https://downloads.dcos.io/dcos/stable/1.10.1/dcos_generate_config.sh
-
-include ./resources/gce.template.mk
-include ./resources/aws.template.mk
-include ./resources/azure.template.mk
+MASTER_IP_FILE := .master_ip
+MASTER_ELB_IP_FILE := .master_elb_ip
+TERRAFORM_INSTALLER_URL := github.com/dcos/terraform-dcos
+DCOS_LAUNCH_VERSION := 0.5.7
```
Comment (pires) [inline Makefile]: Actually, you want the DC/OS CLI (`dcos`) version, not `dcos-launch` which you are deprecating in this PR.

File: Makefile [hunk 2]
```diff
@@ -1,115 +1,98 @@
...
+get-cli:
+	$(eval export DCOS_LAUNCH_VERSION)
+	$(eval export KUBERNETES_VERSION)
+	scripts/get_cli
+
+azure: clean
+	mkdir .deploy
+	cd .deploy; \
+	cp ../resources/desired_cluster_profile.azure desired_cluster_profile; \
+	terraform init -from-module $(TERRAFORM_INSTALLER_URL)//azure
```
Comment (pires) [inline Makefile]: Remove duplicated forward-slash in `//azure`.

File: Makefile [hunk 3]
```diff
@@ -1,115 +1,98 @@
...
+
+aws: clean
+	mkdir .deploy
+	cd .deploy; \
+	cp ../resources/desired_cluster_profile.aws desired_cluster_profile; \
+	terraform init -from-module $(TERRAFORM_INSTALLER_URL)//aws
```
Comment (pires) [inline Makefile]: Remove duplicated forward-slash in `//aws`.

File: Makefile [hunk 4]
```diff
@@ -1,115 +1,98 @@
...
+
+gce: clean
+	$(RM) -r .deploy
+	mkdir .deploy
+	cd .deploy; \
+	cp ../resources/desired_cluster_profile.gce desired_cluster_profile; \
+	terraform init -from-module $(TERRAFORM_INSTALLER_URL)//gcp; \
```
Comment (pires) [inline Makefile]: Remove duplicated forward-slash in `//gcp`.

File: Makefile [hunk 5]
```diff
@@ -1,115 +1,98 @@
...
+	rm desired_cluster_profile.tfvars.example
```
Comment (pires) [inline Makefile]: Why is this here and not on `aws` and `azure`?

File: Makefile [hunk 6]
```diff
@@ -1,115 +1,98 @@
...
+	terraform init -from-module $(TERRAFORM_INSTALLER_URL)//azure
+
+aws: clean
+	mkdir .deploy
+	cd .deploy; \
+	cp ../resources/desired_cluster_profile.aws desired_cluster_profile; \
+	terraform init -from-module $(TERRAFORM_INSTALLER_URL)//aws
+
+gce: clean
+	$(RM) -r .deploy
```
Comment (pires) [inline Makefile]: Why is this here and not on `aws` and `azure`?

Comment (pires) [reply Makefile]: Also, doesn't `clean` take care of removing this?

File: Makefile [hunk 7]
```diff
@@ -1,115 +1,98 @@
...
 define get_master_ip
 $(shell test -f $(MASTER_IP_FILE) || \
-	dcos-launch describe | jq '.["masters"] | .[] | .["public_ip"]' | head -1 | sed s/\"//g | sed s/\:[0-9]*//g > $(MASTER_IP_FILE))
+	terraform output -state=.deploy/terraform.tfstate "Mesos Master Public IP" | head -1 > $(MASTER_IP_FILE))
 $(eval MASTER_IP := $(shell cat $(MASTER_IP_FILE)))
 endef
 
-launch-dcos: $(CLUSTER_INFO_FILE)
-	dcos-launch wait
+get-master-elb-ip:
```
Comment (pires) [inline Makefile]: ELB is unique to AWS. Does this only work for AWS? What about Azure and GCP?

File: Makefile [hunk 8]
```diff
@@ -1,115 +1,98 @@
...
+
+destroy-dcos:
+	$(RM) $(MASTER_IP_FILE)
+	$(RM) $(MASTER_ELB_IP_FILE)
+	cd .deploy; \
+	terraform destroy -var-file desired_cluster_profile
+
+clean:
+	$(RM) -r .deploy
+	$(RM) dcos
```
Comment (pires) [inline Makefile]: What if `dcos` was previously installed?

File: Makefile [hunk 9]
```diff
@@ -1,115 +1,98 @@
...
+	$(RM) kubectl
```
Comment (pires) [inline Makefile]: What if `kubectl` was previously installed?

File: README.md [hunk 10]
```diff
@@ -19,83 +19,115 @@ First, make sure your cluster fulfil the [Kubernetes package default requirement
 
 Then, check the requirements for running this quickstart:
 
-* Google Cloud (GCE) credentials with the necessary [permissions](docs/gce.md)
+* [Terraform 0.11.x](https://www.terraform.io/downloads.html). On MacOS, you can use [brew](https://brew.sh/) for that.
```
Comment (pires) [inline README.md]: Maybe we should run a version check before running any Terraform commands?

File: README.md [hunk 11]
```diff
@@ -19,83 +19,115 @@ First, make sure your cluster fulfil the [Kubernetes package default requirement
...
-
-make destroy-dcos
-# Deletes the DC/OS cluster.
+Install DC/OS cli `dcos` and Kubernetes `kubectl`.
+```
+make get-cli
 ```
 
-## Installing DC/OS CLI
+Files `dcos` and `kubectl` will be downloaded to the current folder, please copy them for example, to `/usr/local/bin/`, or any other folder set in your `path`, so they can be invoked later one by install.
```
Comment (pires) [inline README.md]: If the user does move the binaries elsewhere it means that `make clean` won't actually remove them. This is confusing and somewhat poor UX. I'd rather have the user installing their own way and just make sure the versions are we expect them, just like we're doing with `terraform`.

File: README.md [hunk 12]
```diff
@@ -19,83 +19,115 @@ First, make sure your cluster fulfil the [Kubernetes package default requirement
...
+Install cluster.
+```
+make deploy
+```
 
-Use the Kubernetes command-line tool, kubectl, to deploy and manage applications on Kubernetes. Using kubectl, you can inspect cluster resources; create, delete, and update components; and look at your new cluster and bring up example apps.
+Terraform will be used to set infra on your cloud provider, and then to install DC/OS cluster. When DC/OS is ready Kubernetes cluster will be bootstrapped.
 
-Follow instructions [here](https://kubernetes.io/docs/tasks/tools/install-kubectl/) to download and install.
+Then wait till all Kubernetes packages get installed.
```
Comment (pires) [inline README.md]: There's only one Kubernetes package to be installed so this is misleading.

File: README.md [hunk 13]
```diff
@@ -19,83 +19,115 @@ First, make sure your cluster fulfil the [Kubernetes package default requirement
...
+```
 
-## Installing kubectl
+Install cluster.
+```
+make deploy
+```
 
-Use the Kubernetes command-line tool, kubectl, to deploy and manage applications on Kubernetes. Using kubectl, you can inspect cluster resources; create, delete, and update components; and look at your new cluster and bring up example apps.
+Terraform will be used to set infra on your cloud provider, and then to install DC/OS cluster. When DC/OS is ready Kubernetes cluster will be bootstrapped.
```
Comment (pires) [inline README.md]: Please, replace with something like:
>Terraform will now try and provision the infrastructure on your chosen cloud-provider and then proceed to install DC/OS. When DC/OS is up and running, the Kubernetes package will be installed too.

File: README.md [hunk 14]
```diff
@@ -149,8 +181,13 @@ $ curl http://localhost:9000
...
 ```
 
-We are now ready to install and configure `kubectl`, the Kubernetes CLI tool. For the sake of simplicity, we'll be covering the set-up alone:
-```bash
+We are now ready to configure `kubectl`, the Kubernetes CLI tool.
+```
+make kubectl-config
+```
+
+Which will set cluster `context`.
```
Comment (pires) [inline README.md]: Why is context quoted?

File: docs/aws.md [hunk 15]
```diff
@@ -1,45 +1,69 @@
...
-make uninstall
+Set AWS as cloud provider.
+```
+make aws
+```
+The command above will download necessary [Terraform files](https://github.com/dcos/terraform-dcos/tree/master/aws) to `.deploy` folder.
 
-# Delete the DC/OS cluster.
-make destroy-dcos
+Make updates if you need to (e.g. more private agents) to `.deploy/desired_cluster_profile`, please do not change VMs to lover spec type, as then Kubernetes install will fail.
```
Comment (pires) [inline docs/aws.md]: s/lover/lower

File: Makefile (hunk 5 above)
Comment (rimusz) [reply Makefile]: only gcp `desired_cluster_profile.tfvars.example` has this file

File: Makefile (hunk 7 above)
Comment (rimusz) [reply Makefile]: it is just the naming in TF, it works in azure and gcp

File: Makefile (hunk 9 above)
Comment (rimusz) [reply Makefile]: I clean up the only current folder

File: Makefile (hunk 5 above)
Comment (pires) [reply Makefile]: But why do we need to remove it?

File: Makefile (hunk 7 above)
Comment (pires) [reply Makefile]: OK so please rename the goal

Comment (rimusz) [reply Makefile]: addressed

File: Makefile (hunk 8 above)
Comment (rimusz) [reply Makefile]: addressed not to remove dcos cli

File: Makefile (hunk 9 above)
Comment (rimusz) [reply Makefile]: addressed not to remove kubectl

File: README.md (hunk 11 above)
Comment (rimusz) [reply README.md]: addressed not to remove dcos cli and kubectl

File: README.md (hunk 13 above)
Comment (rimusz) [reply README.md]: addresed

**Comment by rimusz** [comment]
cc @smugcloud

**Comment by sreis** [comment]
Thank you @rimusz! 

I tested this on GCP and with the fix in upstream [terraform-dcos](https://github.com/dcos/terraform-dcos) it worked like a charm.

Now (all) we need is to port it to the dcos-kubernetes repo.

**Comment by pires** [comment]
@sreis a review here would still be nice!

**Comment by rimusz** [comment]
@sreis upstream (terraform-dcos](https://github.com/dcos/terraform-dcos) also needs `IAM Policy and Tag` for dc/os agents so AWS cloud provider support will work with the latest kubernetes framework release. I will git it a stab on monday.

**Comment by pires** [comment]
FYI I'm sharing the load with @rimusz and will be pushing my changes to his branch today or tomorrow.

**Comment by rimusz** [comment]
closing in favour of #50

**Comment by pires** [comment]
@rimusz there was no need to close since #50 already "Closes #49" and when it gets (eventually) merged, Github closes this one automatically ;)

**Comment by rimusz** [comment]
ah, just a bad habit to keep everything clean :)
